### PR TITLE
celery: addition of task to send email via Celery

### LINKED ISSUE #1: celery: add tasks for sending emails via celery
- [ ] Add optional dependency on celery (i.e. add celery require to ``extra_requires``)
- [ ] Add task which receives and email message and sends it.

### PR DESCRIPTION
**Pull request by JavierDelgadoFernandez**
* Added Task module and implemented a method
  which sends an email using Celery. (Closes #1)

Signed-off-by: Javier Delgado <javier.delgado.fernandez@cern.ch>

### CODE REVIEW

**Comment by lnielsen** [comment]
Looks good. Please rebase on top of master.

**Comment by JavierDelgadoFernandez** [comment]
@lnielsen Done.

File: setup.py [hunk 1]
```diff
@@ -45,9 +45,12 @@
 ]
 
 extras_require = {
+    'celery': [
+        'celery>=3.1.0',
+        ],
     'docs': [
         "Sphinx>=1.3",
-    ],
+        ],
```
Comment (jirikuncar) [inline setup.py]: overindented

**Comment by JavierDelgadoFernandez** [comment]
@jirikuncar Overindented fixed.

File: setup.py [hunk 2]
```diff
@@ -61,6 +64,7 @@
 
 install_requires = [
     'Flask-Mail>=0.9.1',
+    'Flask-CeleryExt>=0.1.0'
```
Comment (jirikuncar) [inline setup.py]: isn't in optional together with `celery`?

File: invenio_mail/ext.py [hunk 3]
```diff
@@ -67,5 +67,11 @@ def init_config(self, app):
...
         """Print mail to stream."""
         with self._lock:
-            self.stream.write("{0}\n{1}\n".format(message.as_string(), '-'*79))
-            self.stream.flush()
+            invenio_mail = app.extensions['invenio-mail']
+            invenio_mail.stream.write(
+                "{0}\n{1}\n".format(message.as_string(), '-'*79))
+            invenio_mail.stream.flush()
+
+    def send_email(self, message):
```
Comment (jirikuncar) [inline invenio_mail/ext.py]: IMHO this should not be here. Users of `Invenio-Mail` should decide how they are going to send the email.

File: invenio_mail/ext.py [hunk 4]
```diff
@@ -31,7 +31,7 @@
 
 from flask_mail import Mail, email_dispatched
 
-PY3 = sys.version_info[0] == 3
+from .tasks import send_email
```
Comment (jirikuncar) [inline invenio_mail/ext.py]: It forces celery to be installed. Please remove it from here.

File: invenio_mail/tasks.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+# -*- coding: utf-8 -*-
```
Comment (jirikuncar) [inline invenio_mail/tasks.py]: ping @bruno314 - what is the name for entrypoint that should specify this tasks module?

Comment (bruno314) [inline invenio_mail/tasks.py]: It's ```invenio_celery.tasks```

File: invenio_mail/ext.py [hunk 6]
```diff
@@ -67,5 +67,11 @@ def init_config(self, app):
     def print_email(self, message, app):
         """Print mail to stream."""
         with self._lock:
-            self.stream.write("{0}\n{1}\n".format(message.as_string(), '-'*79))
-            self.stream.flush()
+            invenio_mail = app.extensions['invenio-mail']
```
Comment (jirikuncar) [inline invenio_mail/ext.py]: I don't really follow this change. You are using `self._lock` which is possibly on different `invenio-mail` instance.

a) revert the change
b) create static method and use `invenio_mail._lock` instead of `self._lock`
c) propose different change

Comment (JavierDelgadoFernandez) [inline invenio_mail/ext.py]: It is true that there are several instances of ```invenio-mail```. I think that the easiest change is something like:
```python
invenio_mail = app.extensions['invenio-mail']
with invenio_mail._lock:

Because the instance in app.extensions is always the same.

Comment (jirikuncar) [inline invenio_mail/ext.py]: @JavierDelgadoFernandez it's maybe the easiest but I don't see why did you make this change.

One can always do the same with: `app.extensions['invenio-mail'].print_email(message)`

File: invenio_mail/ext.py [hunk 7]
```diff
@@ -67,5 +67,11 @@ def init_config(self, app):
     def print_email(self, message, app):
```
Comment (jirikuncar) [inline invenio_mail/ext.py]: do we need the `app` argument?

File: invenio_mail/ext.py (hunk 6 above)
Comment (lnielsen) [inline invenio_mail/ext.py]: This is actually bug. I believe the print_email should be taken out of the extension - i.e. it should be something like:

```python
class InvenioMail(object):
    """Invenio-Mail extension.
    Mails are only printed to the stream if ``MAIL_SUPPRESS_SEND`` is ``True``.
    :param app: Flask application object.
    :param stream: Stream to print emails to. Defaults to ``sys.stdout``.
    """

    def __init__(self, app=None, stream=None):
        """Extension initialization."""
        self.stream = stream or sys.stdout
        if app:
            self.init_app(app)

    def init_app(self, app):
        """Flask application initialization."""
        self.init_config(app)
        if 'mail' not in app.extensions:
            Mail(app)
        if app.config.get('MAIL_SUPPRESS_SEND', False) or app.debug:
            email_dispatched.connect(print_email)
        app.extensions['invenio-mail'] = self

    def init_config(self, app):
        """Initialize configuration."""
        app.config.setdefault('MAIL_DEBUG', app.debug)
        app.config.setdefault('MAIL_SUPPRESS_SEND', app.debug or app.testing)


def print_email(message, app):
    """Print mail to stream."""
    stream = app.extensions['invenio-mail'].stream
    stream.write("{0}\n{1}\n".format(message.as_string(), '-'*79))
    stream.flush()
```

File: invenio_mail/ext.py (hunk 3 above)
Comment (lnielsen) [inline invenio_mail/ext.py]: Yes, if you want to use the celery task to send emails, I think the module should explicit import it. Otherwise you simply use Message().send() from Flask-Mail.

**Comment by JavierDelgadoFernandez** [comment]
@jirikuncar I have re-pushed the changes.

File: setup.py [hunk 8]
```diff
@@ -45,6 +45,10 @@
 ]
 
 extras_require = {
+    'celery': [
+        'celery>=3.1.0',
+        'Flask-CeleryExt>=0.1.0'
```
Comment (jirikuncar) [inline setup.py]: please add comma (`,`) at the end of line

**Comment by JavierDelgadoFernandez** [comment]
@jirikuncar Done.

File: tests/conftest.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+            'SQLALCHEMY_DATABASE_URI', 'sqlite://'
+        ),
+        CELERY_ALWAYS_EAGER=True,
+        CELERY_RESULT_BACKEND="cache",
+        CELERY_CACHE_BACKEND="memory",
+        CELERY_EAGER_PROPAGATES_EXCEPTIONS=True,
+        MAIL_SUPPRESS_SEND=True
+    )
+    FlaskCeleryExt(app)
+    fake_stream = FakeStream()
```
Comment (jirikuncar) [inline tests/conftest.py]: now when we solved the issue with `self._lock` can you use `StringIO` instead and test the content too?

File: invenio_mail/ext.py [hunk 10]
```diff
@@ -31,7 +31,15 @@
 
 from flask_mail import Mail, email_dispatched
 
-PY3 = sys.version_info[0] == 3
+
+def print_email(message, app):
+    """Print mail to stream."""
+    # Signal handler for email_dispatched signal.
```
Comment (jirikuncar) [inline invenio_mail/ext.py]: @JavierDelgadoFernandez this was an example. It should be part of the docstring after extending and reformatting.

File: tests/conftest.py (hunk 9 above)
Comment (JavierDelgadoFernandez) [inline tests/conftest.py]: @jirikuncar I have removed the ```FakeStream``` class and the tests are now using the ```StringIO```. There is a new test checking the ```StringIO``` output using Celery.

File: invenio_mail/tasks.py (hunk 5 above)
Comment (jirikuncar) [inline invenio_mail/tasks.py]: please add this entrypoint group

File: invenio_mail/ext.py [hunk 11]
```diff
@@ -31,7 +31,22 @@
 
 from flask_mail import Mail, email_dispatched
 
-PY3 = sys.version_info[0] == 3
+
+def print_email(message, app):
+    """
+    Print mail to stream.
```
Comment (jirikuncar) [inline invenio_mail/ext.py]: Please start docstring on the same line: `"""Print mail to stream.\n\nSignal handler ...`

**Comment by JavierDelgadoFernandez** [comment]
@jirikuncar Done.
